File: 04-Data-Structures/code/python/seb_ds/graph.py
"""Undirected graphs (adjacency list and adjacency matrix). Mirrors `typescript/src/graph.ts`."""

from __future__ import annotations

from typing import Dict, List, Set

from .errors import DSError


class AdjListGraph:
    def __init__(self) -> None:
        self._adj: Dict[str, Set[str]] = {}

    def add_vertex(self, v: str) -> None:
        if v not in self._adj:
            self._adj[v] = set()

    def add_edge(self, u: str, v: str) -> None:
        self.add_vertex(u)
        self.add_vertex(v)
        self._adj[u].add(v)
        self._adj[v].add(u)

    def neighbors(self, v: str) -> List[str]:
        if v not in self._adj:
            raise DSError("missing", f"vertex {v} not found")
        return sorted(self._adj[v])

    def vertex_count(self) -> int:
        return len(self._adj)

    def edge_count(self) -> int:
        total = sum(len(s) for s in self._adj.values())
        return total // 2

    def check_invariants(self) -> None:
        for v, neighbors in self._adj.items():
            for n in neighbors:
                if v not in self._adj.get(n, set()):
                    raise DSError("invalid", "asymmetric adjacency entry")


class AdjMatrixGraph:
    def __init__(self, n: int) -> None:
        if n < 0:
            raise DSError("invalid", "n must be >= 0")
        self._n = n
        self._matrix: List[List[bool]] = [[False] * n for _ in range(n)]

    def add_vertex(self) -> int:
        for row in self._matrix:
            row.append(False)
        self._matrix.append([False] * (self._n + 1))
        self._n += 1
        return self._n - 1

    def _check_vertex(self, v: int) -> None:
        if v < 0 or v >= self._n:
            raise DSError("index", f"vertex {v} out of bounds")

    def add_edge(self, u: int, v: int) -> None:
        self._check_vertex(u)
        self._check_vertex(v)
        self._matrix[u][v] = True
        self._matrix[v][u] = True

    def neighbors(self, v: int) -> List[int]:
        self._check_vertex(v)
        return [i for i in range(self._n) if self._matrix[v][i]]

    def vertex_count(self) -> int:
        return self._n

    def edge_count(self) -> int:
        count = 0
        for i in range(self._n):
            for j in range(i + 1, self._n):
                if self._matrix[i][j]:
                    count += 1
        return count

    def check_invariants(self) -> None:
        for i in range(self._n):
            if len(self._matrix[i]) != self._n:
                raise DSError("invalid", "matrix row length mismatch")
            for j in range(self._n):
                if self._matrix[i][j] != self._matrix[j][i]:
                    raise DSError("invalid", "asymmetric matrix entry")

File: 04-Data-Structures/code/typescript/src/graph.ts
import { DSError } from "./errors.js";

/** Undirected graph over string-labeled vertices, backed by adjacency sets. */
export class AdjListGraph {
  private adj: Map<string, Set<string>> = new Map();

  addVertex(v: string): void {
    if (!this.adj.has(v)) this.adj.set(v, new Set());
  }

  addEdge(u: string, v: string): void {
    this.addVertex(u);
    this.addVertex(v);
    this.adj.get(u)!.add(v);
    this.adj.get(v)!.add(u);
  }

  neighbors(v: string): string[] {
    const set = this.adj.get(v);
    if (!set) throw new DSError("missing", `vertex ${v} not found`);
    return Array.from(set).sort();
  }

  vertexCount(): number {
    return this.adj.size;
  }

  edgeCount(): number {
    let sum = 0;
    for (const set of this.adj.values()) sum += set.size;
    return sum / 2;
  }

  checkInvariants(): void {
    for (const [v, neighbors] of this.adj) {
      for (const n of neighbors) {
        if (!this.adj.get(n)?.has(v)) throw new DSError("invalid", "asymmetric adjacency entry");
      }
    }
  }
}

/** Undirected graph over vertices 0..n-1, backed by a boolean adjacency matrix. */
export class AdjMatrixGraph {
  private matrix: boolean[][];
  private n: number;

  constructor(n: number) {
    if (n < 0) throw new DSError("invalid", "n must be >= 0");
    this.n = n;
    this.matrix = Array.from({ length: n }, () => new Array<boolean>(n).fill(false));
  }

  addVertex(): number {
    for (const row of this.matrix) row.push(false);
    this.matrix.push(new Array<boolean>(this.n + 1).fill(false));
    this.n++;
    return this.n - 1;
  }

  private checkVertex(v: number): void {
    if (v < 0 || v >= this.n) throw new DSError("index", `vertex ${v} out of bounds`);
  }

  addEdge(u: number, v: number): void {
    this.checkVertex(u);
    this.checkVertex(v);
    this.matrix[u][v] = true;
    this.matrix[v][u] = true;
  }

  neighbors(v: number): number[] {
    this.checkVertex(v);
    const out: number[] = [];
    for (let i = 0; i < this.n; i++) if (this.matrix[v][i]) out.push(i);
    return out;
  }

  vertexCount(): number {
    return this.n;
  }

  edgeCount(): number {
    let count = 0;
    for (let i = 0; i < this.n; i++) for (let j = i + 1; j < this.n; j++) if (this.matrix[i][j]) count++;
    return count;
  }

  checkInvariants(): void {
    for (let i = 0; i < this.n; i++) {
      if (this.matrix[i].length !== this.n) throw new DSError("invalid", "matrix row length mismatch");
      for (let j = 0; j < this.n; j++) {
        if (this.matrix[i][j] !== this.matrix[j][i]) throw new DSError("invalid", "asymmetric matrix entry");
      }
    }
  }
}
